[Issue facebook/css-layout#78]: Implemented alignContent ;

prenaux opened the pull request:
So this implements align-content. It follows the W3C specs, that's how it works in Chrome as far as I have tested. I tried to make sure that it didn't touch any pre-existing code as much as possible.

All the current tests are working (in RunLayoutTests.html and with make - JS, C & Java). RunLayoutRandomTests.html is failing but then it was failing without these changes so I suppose its an issue with the random test suite not being up-to-date.

I'm going to add some more test cases in Jasmine and add this in react-native. From simple tests it works fine in react-native but I need to add alignContent in there to make sure that the new feature works correctly.

Let me know if you have any comment about the implementation itself.

vjeux commented:
Thanks a lot for working on this! I'm going to take a look at it shortly :)

vjeux commented:
Right now random tests are broken for some really edgy cases. What I usually do is to disable the attributes that are making those tests to fail until they are clean. Then, add your new feature and run the automated tests.

I cannot really bring this in without any tests, can you make sure to add at least 5 test cases to make sure that your feature is working correctly, but more importantly that if we make further modifications down the road, it won't break down.

Thanks! This is super exciting!

prenaux commented:
Yes, the additional test cases will be there shortly.

vjeux commented:
Awesome thanks :)

prenaux commented:
I've added 16 test cases for the new alignContent code.

I also added pixel snapping in the layout tests, please check the code for a more complete description, but the issue is that chrome seems to change its layout code so the floating points values of layoutDom change from version to version, the snapping code I added assumes that it's ok to be 'only' pixel perfect, and that we don't care if there's a sub-pixel difference.

I have Chrome 42.0.2311.135 (64-bit) on OSX and 19 tests fail (without the alignContent changes - just after pulling the current master) because of floating point deviations. The biggest issue imo is that my Windows machine still works fine, so even if the code is 'fixed' for Chrome 42 on OSX other platforms might still give different results.

prenaux commented:
Added alignContent in the test or random permutations, and also added the pixel snapping in the random tests - so that sub-pixel errors are not counted as 'random' errors.

vjeux commented on src/Layout-test-utils.js:
```diff
@@ -227,6 +240,27 @@ var layoutTestUtils = (function() {
+      var val = aObj[key];
+      switch (typeof(val)) {
+      case 'number': {
+        aObj[key] = Math.floor((val*testMeasurePrecision)+0.5)/testMeasurePrecision;
```
nit: can you put space around *, + and /

vjeux commented on src/Layout-test-utils.js:
```diff
@@ -227,6 +240,27 @@ var layoutTestUtils = (function() {
+      if (!aObj.hasOwnProperty(key))
+        continue;
+      var val = aObj[key];
+      switch (typeof(val)) {
```
nit: `typeof val`

vjeux replied:
Can you use an if/then/else, it makes the code shorter:

```
if (typeof val === 'number') {
  obj[key] = ...;
} else if (typeof val === 'object') {
  inplaceRondNumbersInObject(val);
}
```

vjeux commented on src/Layout-test-utils.js:
```diff
@@ -227,6 +240,27 @@ var layoutTestUtils = (function() {
+      return;
+
+    for (var key in aObj) {
+      if (!aObj.hasOwnProperty(key))
```
nit: please always add {}

vjeux commented on src/Layout-test-utils.js:
```diff
@@ -227,6 +240,27 @@ var layoutTestUtils = (function() {
     return layout;
   }
 
+  function inplaceRoundNumbersInObject(aObj) {
```
Can you name it `obj`

vjeux commented on src/Layout-test-utils.js:
```diff
@@ -227,6 +240,27 @@ var layoutTestUtils = (function() {
   }
 
+  function inplaceRoundNumbersInObject(aObj) {
+    if (!testMeasurePrecision) // undefined/0, disables rounding
```
braces

vjeux commented on src/Layout-test-utils.js:
```diff
@@ -337,11 +371,13 @@ var layoutTestUtils = (function() {
     div.style.display = 'flex';
     div.style.flexDirection = 'column';
     div.style.alignItems = 'flex-start';
+    div.style.alignContent = 'stretch';
```
didn't you say that we currently have `flex-start` as the default?

prenaux replied:
'flex-start' is similar to the previous React 'result'... but 'stretch' is the default in standard CSS so I left it at that.

vjeux replied:
I found flexbox defaults to be pretty terrible, the ones that are in React native work really well to build apps. Also, when I import this back to fb codebase, I don't want everything to break under me!

prenaux replied:
Ok, fine by me :) I'll change it to flex-start.

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
 
+  // <Loop DD>
+  //
+  // PIERRE: More than one line, we need to layout the crossAxis according to
```
Can you use the following formatting:

```
// Note(prenaux): 
```

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+        }
+        else if (alignItem == CSS_ALIGN_STRETCH) {
+          crossPosition = currentLead + getMargin(child,leading[crossAxis]);
+          // TODO: Correctly set the height of items with undefined (auto)
```
TODO(prenaux)

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
     startLine = endLine;
   }
 
+  // <Loop DD>
```
Loop E?

prenaux replied:
There's loop E below (line 794), I can rename that to F, and put this to E.

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+    }
+    else if (alignContent == CSS_ALIGN_STRETCH) {
+      if (nodeCrossAxisInnerSize > linesCrossDim) {
+        crossDimAdd = (remainingCrossDim / linesCount);
```
nit: no need for parenthesis

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+      int lineIndex = -1;
+
+      // get the first child on the current line
+      {
```
why do you create a new block?

prenaux replied:
I was declaring another variable in there, I switched to use the child variable when I converted the code to JS/transpiler. Removed it.

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+    if (alignContent == CSS_ALIGN_FLEX_END) {
+      currentLead += remainingCrossDim;
+    }
+    else if (alignContent == CSS_ALIGN_CENTER) {
```
nit: `}` and `else if` on the same line

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+    }
+    else if (alignContent == CSS_ALIGN_CENTER) {
+      currentLead += remainingCrossDim / 2;
+    }
```
same here

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+          break;
+        }
+        if (!isUndefined(child->layout.dimensions[dim[crossAxis]])) {
+          lineHeight = fmaxf(lineHeight,child->layout.dimensions[dim[crossAxis]] +
```
nit: indent like this:

```
lineHeight = fmaxf(
  lineHeight,child->layout.dimensions[dim[crossAxis]] +
  getMarginAxis(child,crossAxis)
);
```

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+        if (alignItem == CSS_ALIGN_FLEX_START) {
+          crossPosition = currentLead + getMargin(child,leading[crossAxis]);
+        }
+        else if (alignItem == CSS_ALIGN_FLEX_END) {
```
same line

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+            getMargin(child,trailing[crossAxis]) -
+            child->layout.dimensions[dim[crossAxis]];
+        }
+        else if (alignItem == CSS_ALIGN_CENTER) {
```
same line

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+          float childHeight = child->layout.dimensions[dim[crossAxis]];
+          crossPosition = currentLead + ((lineHeight - childHeight)/2);
+        }
+        else if (alignItem == CSS_ALIGN_STRETCH) {
```
same line

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+          crossPosition = currentLead + ((lineHeight - childHeight)/2);
+        }
+        else if (alignItem == CSS_ALIGN_STRETCH) {
+          crossPosition = currentLead + getMargin(child,leading[crossAxis]);
```
nit: space after `,`

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+        css_align_t alignItem = getAlignItem(node, child);
+        float crossPosition = child->layout.position[pos[crossAxis]]; // preserve current position if someting goes wrong with alignItem?
+        if (alignItem == CSS_ALIGN_FLEX_START) {
+          crossPosition = currentLead + getMargin(child,leading[crossAxis]);
```
nit: space after `,`

vjeux commented on src/__tests__/Layout-test.js:
```diff
@@ -1586,5 +1587,78 @@ describe('Layout', function() {
 
+});
+
+describe('Layout alignContent', function() {
```
Can you figure out why those tests are not being transpiled to C and Java? Is it because you are creating a new describe?

prenaux replied:
Most likely, going to check that, I didn't know they were supposed to be transpiled to C ;)

vjeux commented on src/__tests__/Layout-test.js:
```diff
@@ -1586,5 +1587,78 @@ describe('Layout', function() {
+  var alignContentData = {
+    style: {width: 300, height: 380, flexDirection: 'row', flexWrap: 'wrap'},
+    children: [
+      /* 0 */  { style: {width: 50, height: 50, margin: 10} },
```
nit: no space right after `{` and before the `}`

vjeux commented on src/__tests__/Layout-test.js:
```diff
@@ -1586,5 +1587,78 @@ describe('Layout', function() {
+    testLayout(
+      alignContentData,
+      { width:300,height:380,top:0,left:0, children:[
+        {width:50,height:50,top:10,left:10},
```
nit: space after `:` and `,`

vjeux commented on src/__tests__/Layout-test.js:
```diff
@@ -1586,5 +1587,78 @@ describe('Layout', function() {
+  it('should layout with alignContent: stretch, and alignItems: flex-start', function() {
+    testLayout(
+      alignContentData,
+      { width:300,height:380,top:0,left:0, children:[
```
nit: no space after `{` but space before `[`

vjeux commented on src/__tests__/Layout-test.js:
```diff
@@ -1586,5 +1587,78 @@ describe('Layout', function() {
+  });
+
+  function testAlignContent(aAlignContent, aAlignItems) {
+    it('should layout with alignContent: '+aAlignContent+', and alignItems: '+aAlignItems, function() {
```
nit: space around `+`

vjeux commented on src/__tests__/Layout-test.js:
```diff
@@ -1586,5 +1587,78 @@ describe('Layout', function() {
+    });
+  }
+
+  // testAlignContent('stretch', 'flex-start'); // above with expected value data
```
why is it commented out?

prenaux replied:
Its the same as 'should layout with alignContent: stretch, and alignItems: flex-start'. Difference is that those extra 15 tests are only against the DOM, they don't have a set of expected data in the code... mostly because I didn't really want to copy/paste the data 15 times from Chrome. I didn't know this was transpiled to C eventually, probably going to have to update it ;)

vjeux commented on src/__tests__/Layout-test.js:
```diff
@@ -1586,5 +1587,78 @@ describe('Layout', function() {
+    );
+  });
+
+  function testAlignContent(aAlignContent, aAlignItems) {
```
nit: remove `a` at the front of the variable name

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+      }
+
+      // find the first node on the first line
+      for (i = 0; i < node.children.length; ) {
```
this loop is super scary, it is hard to know if it's going to terminate by reading the code. Is there any way you can make it a standard loop over the array?

vjeux replied:
what about for

```javascript
for (lineIndex = 0; lineIndex < linesCount; ++lineIndex) {
```

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+          }
+
+          var/*css_align_t*/ alignItem = getAlignItem(node, child);
+          var/*float*/ crossPosition = child.layout[pos[crossAxis]]; // preserve current position if someting goes wrong with alignItem?
```
you don't need this since you handle all the cases

prenaux replied:
My defensive reflexes from C, its in case someone adds another possible value for alignItem that I'm not handling.

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+        }
+        else if (alignItem == CSS_ALIGN_CENTER) {
+          float childHeight = child->layout.dimensions[dim[crossAxis]];
+          crossPosition = currentLead + ((lineHeight - childHeight)/2);
```
nit: you don't need parenthesis around (a / 2) because / has precedence over +
nit: space around /

vjeux commented on src/Layout.c:
```diff
@@ -770,9 +777,112 @@ static void layoutNodeImpl(css_node_t *node, float parentMaxWidth) {
+  //
+  if (linesCount > 1 &&
+      (!isUndefined(node->layout.dimensions[dim[crossAxis]])))
+  {
```
nit: don't need parenthesis around !isUndefined and put { on the same line

```
if (linesCount > 1 &&
    !isUndefined(node->layout.dimensions[dim[crossAxis]])) {
```

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+          if (child.lineIndex != lineIndex) {
+            break;
+          }
+          if (!isUndefined(child.layout[dim[crossAxis]])) {
```
what if none of the children have a specified dimension? Is it properly handled?

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+        var/*int*/ startIndex = i;
+        var/*int*/ lineIndex = -1;
+
+        // get the first child on the current line
```
if you use my suggestion above, you can just get rid of this block to find the lineIndex since you have it

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+          }
+        }
+        var/*int*/ endIndex = ii;
+        lineHeight += crossDimAdd;
```
can you make crossDimAdd variable name more descriptive?

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+    // PIERRE: More than one line, we need to layout the crossAxis according to
+    // alignContent.
+    //
+    // Note that we could probably remove <Loop D> and handle the one line case
```
I think it's good to have it as a separate loop

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+        var/*float*/ lineHeight = 0;
+        for (ii = startIndex; ii < node.children.length; ++ii) {
+          child = node.children[ii];
+          if (getPositionType(child) != CSS_POSITION_RELATIVE) {
```
please use !==

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+          if (getPositionType(child) != CSS_POSITION_RELATIVE) {
+            continue;
+          }
+          if (child.lineIndex != lineIndex) {
```
!==

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+
+        for (ii = startIndex; ii < endIndex; ++ii) {
+          child = node.children[ii];
+          if (getPositionType(child) != CSS_POSITION_RELATIVE) {
```
!==

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+          if (alignItem == CSS_ALIGN_FLEX_START) {
+            crossPosition = currentLead + getMargin(child,leading[crossAxis]);
+          }
+          else if (alignItem == CSS_ALIGN_FLEX_END) {
```
===

vjeux commented on src/Layout.js:
```diff
@@ -654,9 +663,112 @@ var computeLayout = (function() {
+            // TODO: Correctly set the height of items with undefined (auto)
+            //       crossAxis dimension.
+          }
+          child.layout[pos[crossAxis]] = crossPosition;
```
since you always assign to child.layout[pos[crossAxis]], do you really need the variable? Can you just assign it in every branch of the if/

prenaux replied:
yup

vjeux commented on src/__tests__/Layout-random-test.js:
```diff
@@ -65,6 +65,7 @@ describe('Random layout', function() {
     randEnum(node, 0.5, 'justifyContent', ['flex-start', 'center', 'flex-end', 'space-between', 'space-around']);
     randEnum(node, 0.5, 'alignItems', ['flex-start', 'center', 'flex-end', 'stretch']);
     randEnum(node, 0.5, 'alignSelf', ['flex-start', 'center', 'flex-end', 'stretch']);
+    randEnum(node, 0.5, 'alignContent', ['flex-start', 'center', 'flex-end', 'stretch']);
```
:+1:

vjeux commented:
Thanks! I added a lot of stylistic comments, but overall the code looks good. I'm excited to ship it :)

prenaux commented:
Ok, I think all has been styled / fixed, let me know if I missed anything :)

vjeux commented:
Thanks i'll take a look sometimes today!

prenaux commented:
Rebased on the latest version.

vjeux commented:
Ergg, I'm sorry but can you rebase, it is no longer merging cleanly. Sorry for the long delay, I marked the mail as read and forgot about it, sorry :(

prenaux commented:
Ok, merged. Re-run the automated tests and the tests I have in React native ;

vjeux commented:
Thanks!
